### ISSUE TITLE
False positive from matching wrong type (`type[Protocol]` rather than `type[_ProtocolMeta]`).

### ISSUE THREAD

**Issue by emcd** [microsoft/pyright#10324]
**Describe the bug**

In the reproducer below, I get the following error:
```text
/home/me/src/python-frigid/bugs/pyright-mcs-factory.py
  /home/me/src/python-frigid/bugs/pyright-mcs-factory.py:30:44 - error: Argument of type "type[type[Protocol]]" cannot be assigned to parameter "clscls_base" of type "type[T@produce_class_factory_class]" in function "produce_class_factory_class"
    Type "type[Protocol]" is not assignable to type "type" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

The type being passed to the argument in question is actually `typing._ProtocolMeta`, which is a subclass of `type`. In light of this, the Pyright diagnostic complaining about `type[ Protocol ]` does not seem correct, since `type[ Protocol ]` is not the same as `type( Protocol )`.

(If I remove the bound on `T`, then Pyright complains about `__new__`, because it cannot infer, as expected, that it is working with a metaclass. And, if I do not qualify `type` with `T` for the argument, then there are other Pyright issues, also as expected.)

**Code or Screenshots**

Reproducer:

```python
from __future__ import annotations

import typing_extensions as typx


T = typx.TypeVar( 'T', bound = type )


def produce_class_factory_class( clscls_base: type[ T ] ) -> type:

    class Class_( clscls_base ):

        def __new__(
            clscls: type[ Class_ ],
            name: str,
            bases: tuple[ type, ... ],
            namespace: dict[ str, typx.Any ],
            **args: typx.Any,
        ) -> Class_:
            return super( ).__new__( clscls, name, bases, namespace, **args )

    return Class_


# No Pyright error.
Mcs = produce_class_factory_class( type )


# Pyright error.
ProtocolMcs = produce_class_factory_class( type( typx.Protocol ) )


# No Pyright error.
class OtherProtocolMcs( type( typx.Protocol ) ):

    def __new__(
        clscls: type[ OtherProtocolMcs ],
        name: str,
        bases: tuple[ type, ... ],
        namespace: dict[ str, typx.Any ],
        **args: typx.Any,
    ) -> OtherProtocolMcs:
        return super( ).__new__( clscls, name, bases, namespace, **args )


# Correct runtime behavior in both CPython and PyPy.

class C( metaclass = Mcs ): pass
print( type( C ) )
print( type( C ).__mro__ )

class PC( typx.Protocol, metaclass = ProtocolMcs ): pass
print( type( PC ) )
print( type( PC ).__mro__ )

class OPC( typx.Protocol, metaclass = OtherProtocolMcs ): pass
print( type( OPC ) )
print( type( OPC ).__mro__ )
```

Output (demonstrating that Python correctly interprets the code):
```text
<class '__main__.produce_class_factory_class.<locals>.Class_'>
(<class '__main__.produce_class_factory_class.<locals>.Class_'>, <class 'type'>, <class 'object'>)
<class '__main__.produce_class_factory_class.<locals>.Class_'>
(<class '__main__.produce_class_factory_class.<locals>.Class_'>, <class 'typing_extensions._ProtocolMeta'>, <class 'typing._ProtocolMeta'>, <class 'abc.ABCMeta'>, <class 'type'>, <class 'object'>)
<class '__main__.OtherProtocolMcs'>
(<class '__main__.OtherProtocolMcs'>, <class 'typing_extensions._ProtocolMeta'>, <class 'typing._ProtocolMeta'>, <class 'abc.ABCMeta'>, <class 'type'>, <class 'object'>)
```

**VS Code extension or command-line**
$ hatch --env develop run pyright --version
pyright 1.1.399

**Comment by erictraut** [comment]
I think pyright's behavior here is justified, so I don't consider this a bug.

The symbol `typing.Protocol` is a special form, and its internal implementation details are not documented, nor are they defined in the typeshed stubs that type checkers rely upon for type information.

Your code is relying on internal types (like `_ProtocolMeta`) and internal runtime implementation details.

The typing spec indicates how `Protocol` should be used in a class definition to define a protocol class, but it doesn't provide any guidance for how `Protocol` should be treated when used in a value expression as you're doing in this code. I don't think I've ever seen it used in a value expression. What you're doing here is highly unusual. Not surprisingly, mypy also emits errors for this code.

I don't plan to add any special casing to handle this use case unless / until the intended behavior is documented in the official Python documentation or in the typing spec.

If you want this to type check, you'll need to apply some workarounds like `cast` or `# type: ignore`.

**Comment by emcd** [comment]
> The symbol `typing.Protocol` is a special form, and its internal implementation details are not documented, nor are they defined in the typeshed stubs that type checkers rely upon for type information.

@erictraut : While it is a special form, it is also part of the Python type system. It is a real class and its metaclass _inherits_ from `abc.ABCMeta`.

> Your code is relying on internal types (like `_ProtocolMeta`) and internal runtime implementation details.

My code is not directly relying on an internal. I deliberately use `type( Protocol )` to avoid relying on an internal. If `typing._ProtocolMeta` is replaced by something else, my code should not be affected. (And whatever it is replaced with will still have to be a metaclass, because `Protocol` must be a class.)

> The typing spec indicates how `Protocol` should be used in a class definition to define a protocol class, but it doesn't provide any guidance for how `Protocol` should be treated when used in a value expression as you're doing in this code. I don't think I've ever seen it used in a value expression. What you're doing here is highly unusual.

Unusual, yes, but that does not mean it is unjustified. The use case for which I am developing this is to reduce code duplication when creating metaclasses to produce classes whose class attributes are immutable. Currently, I have seven different metaclasses with essentially identical implementations, but different bases to avoid metaclass base conflicts. This collapses them down to fewer implementations.

>Not surprisingly, mypy also emits errors for this code.

Mypy emits errors for a lot of things that aren't errors, which is why I rely on Pyright instead.

> If you want this to type check, you'll need to apply some workarounds like `cast` or `# type: ignore`.

I have a `# pyright: ignore` on the real use case, but it seemed worth reporting....
